WireGuardTunnelExtension/PacketTunnelProvider.swift

```diff
@@ -16,53 +15,100 @@ enum PacketTunnelProviderError: Error {
         }
     }()
 
     override func startTunnel(options: [String: NSObject]?, completionHandler: @escaping (Error?) -> Void) {
-        NSLog("WireGuard-eduVPN: Starting tunnel")
-
         guard let protocolConfiguration = self.protocolConfiguration as? NETunnelProviderProtocol,
               let providerConfiguration = protocolConfiguration.providerConfiguration,
-              let wgQuickConfig = providerConfiguration["WireGuardConfig"] as? String else {
-            NSLog("WireGuard-eduVPN: WireGuard config not found")
+              let wgQuickConfig = providerConfiguration[WireGuardProviderConfigurationKeys.wireGuardConfig.rawValue] as? String,
+              let appGroup = providerConfiguration[WireGuardProviderConfigurationKeys.appGroup.rawValue] as? String else {
+            NSLog("Invalid provider configuration for the WireGuard tunnel")
             completionHandler(PacketTunnelProviderError.savedProtocolConfigurationIsInvalid)
             return
         }
 
+        let logger = Logger(appGroup: appGroup, separator: "--- EOF ---", logFileName: "debug.log")
+        self.logger = logger
+
+        logger.log("Starting WireGuard tunnel")
+
         guard let tunnelConfiguration = try? TunnelConfiguration(fromWgQuickConfig: wgQuickConfig) else {
-            NSLog("WireGuard-eduVPN: WireGuard config not parseable")
+            logger.log("WireGuard config not parseable")
             completionHandler(PacketTunnelProviderError.savedProtocolConfigurationIsInvalid)
             return
         }
+        self.tunnelConfiguration = tunnelConfiguration
 
         adapter.start(tunnelConfiguration: tunnelConfiguration) { adapterError in
             if let adapterError = adapterError {
-                NSLog("WireGuard-eduVPN: WireGuard adapter error: \(adapterError.localizedDescription)")
+                logger.log("WireGuard adapter error: \(adapterError.localizedDescription)")
             } else {
                 let interfaceName = self.adapter.interfaceName ?? "unknown"
-                NSLog("WireGuard-eduVPN: Tunnel interface is %@", interfaceName)
+                logger.log("Tunnel interface is \(interfaceName)")
             }
             completionHandler(adapterError)
         }
     }
 
     override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
-        NSLog("WireGuard-eduVPN: Stopping tunnel")
+        logger?.log("Stopping tunnel")
+
         adapter.stop { error in
             if let error = error {
-                NSLog("WireGuard-eduVPN: Failed to stop WireGuard adapter: \(error.localizedDescription)")
+                NSLog("Failed to stop WireGuard adapter: \(error.localizedDescription)")
             }
+            self.logger?.flushToDisk()
             completionHandler()
 
             #if os(macOS)
             // HACK: We have to kill the tunnel process ourselves because of a macOS bug
-            exit(0)
+            // exit(0)
```

Comment:
Since this is commented out, the hack is no longer needed?

Reply:
The hack is still very much required. I'd commented it out while testing the log flushing code. Thanks a lot for spotting this - I was wondering why my disconnects were taking too long.

Reply:
Fixed in comment ac23d90.